### ISSUE TITLE
Events configured "when deleting a row" are not triggered

### ISSUE THREAD

**Issue by gianlucafiore** [saltcorn/saltcorn#2505]
No prewritten action is triggered when configured to be invoked when deleting a record from the declared table

![image](https://github.com/saltcorn/saltcorn/assets/50465150/5bc99763-6996-4325-ab6b-00a9ae818338)

I have these two actions written down that do the same thing, they're just called on delete or insert. because they should capture the actions performed by means of "Edit Badge", I take the insert well, but not the delete

**Comment by gianlucafiore** [comment]
@glutamate

**Comment by gianlucafiore** [comment]
Dear @glutamate Is there any alternative to solve this? The truth is that it is quite dangerous for me, since I have to look for strange gadgets everywhere to be able to solve this problem.

**Comment by glutamate** [comment]
@christianhugoch would you be able to have a look? The problem is that Edit Badge deletes rows using sql instead of `Table.deleteRows` so the event triggers are never run. All we need to do is change the sql query into a call to Table.deleteRows.

**Comment by christianhugoch** [comment]
Yes, I can take a look

**Comment by christianhugoch** [comment]
added a pr. I hope all looks good

The sqlite part is untested, I think at first, we have to make the plugin sqlite ready, before it matters

**Comment by gianlucafiore** [comment]
@christianhugoch  I tried the modifications made for the PR, but honestly it didn't have any effect :(

the event is still not called by the trigger (Delete row)

Thank you both for your attention.

**Comment by christianhugoch** [comment]
Okay, thanks for testing

Are you sure the delete itself worked? There was a permission problem that could cause it to fail.

I added another commit for this, can you update and re-check?

If it still fails, can you give me some insights on the actions and the gui structure? Are the run_js_code actions serverside? I believe client side won't work here, but also not for insert.

**Comment by gianlucafiore** [comment]
I'll tell you from the beginning how what I'm doing works, and what the problems are that I have.

I'm building a module in the system that allows you to make a budget for a project.
This budget is made up of descriptive items, with quantities, unit values, percentage taxes to calculate, and the total calculation of said taxes.

![image](https://github.com/user-attachments/assets/9c983aa2-c3f7-41b1-9fcb-5e8bf4e26335)

In the "Agr. Taxes" column I use the "edit badge" plugin to assign, through a many-to-many relationship, the required taxes, which can be 1 or 2 or 3. It depends on the product being quoted.
When adding said tax, I internally created a script that is called "When inserting a row" and "When deleting a row" that adds the percentages of the taxes assigned to the row being quoted, multiplies the sum by the unit value, by the quantity, the values ​​are stored in fields, and then at the end of everything in the view, aggregations are made to add all the items in the quote, add all the taxes, and add the final totals. (Σ Taxes * v_unit * quantity)


The problems I find with these are:
1-One of the least serious: is that when adding or deleting a badge, the page does not reload, nor does it recalculate the totals visually (Until the page is reloaded manually)

2- The most tedious in question is that when deleting a badge, not only does it not perform the action mentioned above, but it does not call the event that recalculates the totals, and therefore the totals are in disagreement with the taxes it says it has. (They only pay attention to the trigger "when inserted" and not "when deleted"


Here, script of event to execute on insert or delete:

```
let tabla_impuestos = await Table.findOne('impuestos_a_items_presupuesto'); //con tabla original
//let tabla_impuestos = await Table.findOne('impuestos_items_presupuesto');
let tabla_items = await Table.findOne('items_presupuesto');
 
let suma = await tabla_impuestos.aggregationQuery({ 
      nm: {
        field: 'valor_impuesto',
  		aggregate:'sum'
      }
},{
where: {item_presupuesto: item_presupuesto}
}) 
console.log(suma)

let item = await tabla_items.getRow({id:item_presupuesto}) 

await tabla_items.updateRow({impuestos:suma.nm * Number(item.cantidad) * Number(item.valor_unitario)},item_presupuesto);
```

![image](https://github.com/user-attachments/assets/d801bdf4-b130-4250-9d03-6bee24e24a11)

**Comment by gianlucafiore** [comment]
Sure, I actually did a FORK, changed the plugin repo name in the package.json and installed it manually, from the package manager.

**Comment by christianhugoch** [comment]
Hey, just looked at your fork and saw this on main
https://github.com/gianlucafiore/badges/blob/8ca9389ecd88ecc650be66ad144a509fbfe4701d/edit.js#L328

I believe that's the old code, my changes are on a branch (edit-badge-action). 

But I would feel better if we just merge and release it the normal way.

**Comment by glutamate** [comment]
merged and published as 0.2.4

**Comment by gianlucafiore** [comment]
thanks guys, I´will test this. regards!

**Comment by gianlucafiore** [comment]
![image](https://github.com/user-attachments/assets/303f1797-bcc9-4206-a494-4ae5f4a79267)

I definitely think the problem is that the trigger is not executed correctly. I tried creating a "list" with delete buttons for those items, to keep the badge plugin out of the testing, and the event is never called.

**Comment by gianlucafiore** [comment]
yes of course, I will sending to you. Thanks for agility

**Comment by gianlucafiore** [comment]
sended for drive

**Comment by christianhugoch** [comment]
thanks, will take a look soon

**Comment by christianhugoch** [comment]
I believe the problem is that the trigger runs before the actual delete and when you grep into the database you still see the old state.

Can you verify that? I noticed that by deleting multiple items (5-6) and refreshed after each delete.

**Comment by gianlucafiore** [comment]
The truth is that I'm already dizzy with this case, but how did you do the deletion? From the "items_presupuesto" list through "edit_badge"?

I'm going to try it anyway to see if I can detect it that way.

**Comment by gianlucafiore** [comment]
Hello, good morning, @christianhugoch  , I really couldn't detect any change for the better, nor find the problem that you described to me.

**Comment by christianhugoch** [comment]
I'm going to send you screenshots later today. I've resetted the schema but I'm going to set it up again.

TL;DR
I used the badge delete button and saw that the trigger ran. But I believe when you call aggregationQuery, then the row you want to delete is still in the db, this explains why you don't see a change on the gui. Are you really sure this dosen't happen on your machine?

**Comment by christianhugoch** [comment]
This is the initial state:
![initial](https://github.com/user-attachments/assets/ff2e142c-eb25-45db-94f2-d47f23fa8cd4)

Then I deleted the item with impuesto one and the GUI dose not update correctly:
![delete_one](https://github.com/user-attachments/assets/de316a7e-b7e8-4d4b-9cab-3611d70323b5)

But I've added a console.log() on top of the del trigger and you see it in the logs:
![logs_for_delete_one](https://github.com/user-attachments/assets/ad5adb5a-bac5-4d66-afc2-f35edbca4953)

After that I deleted the two items with impuestor two. After first delete:
![second_delete](https://github.com/user-attachments/assets/da11b778-424a-47ba-a5bb-c174fd9dea31)

Nothing happens, as expected. But after the second delete you see an update:
![third_delete](https://github.com/user-attachments/assets/0a4167f0-233d-423f-9c97-48c6d7262a2c)

I'm convinced the update you see in the last delete is based on the first delete, but it ignores the current delete, because the trigger runs before the db operation.

Does this help?

**Comment by gianlucafiore** [comment]
Excellent, I think that is where the problem lies, and what would be the solution or temporary solution?

**Comment by christianhugoch** [comment]
I think we have two options:

1 .In the delete trigger, you have access to the row variable, that's the row you're about to delete. Can you change your logic, to take this value out of the calculation?
2. We could add an option to run delete trigger before or after the db operation.

I think as a temporary fix, I would go with 1.

**Comment by gianlucafiore** [comment]
Thank you very much my friend, I was able to solve it for the moment, thanks to your help with the testing as well...
![image](https://github.com/user-attachments/assets/d4e72575-9e23-46de-923e-0379bc86bb3c)
I had to add that line of code so that the DB is updated by subtracting the previously entered value.

**Comment by christianhugoch** [comment]
Hi, I've looked at it, but I believe with the current code, there is no "easy" way to reload another embedded view or the page after the edit (will think more about it).

There are ways to implement that, but can we put this in another issue?

**Comment by gianlucafiore** [comment]
Ok, of course!
